### PR TITLE
Run Solver from GUI

### PR DESCRIPTION
**Pull request by Necktschnagge**
### Make use of the solver from Qt GUI
- [x] Start Solver ➡️ explores partial state graph
- [x] Obtain all color-aware move paths, classified by their goal state
- [x] sub-classify move paths using transitive closure of interleaving neighbour equivalence
- [x] sorting of paths which remain in the same equivalence class by using some prettiness sorting...
- [x] displaying a representant for each equivalence class in the ListView
- [x] allow user to manually step-by-step play all solver solutions

### Enhancement + Fixes
- [ ] ensure state safety for `GuiInteractiveController` and `GameController` (moved to #82 )

### CODE REVIEW

**Comment by Necktschnagge** [comment]
I merged #64 into this branch.
I'll do a rebase, once #64 is merged into main.

**Comment by Necktschnagge** [comment]
rebased all commits since #64 has been merged

**Comment by Necktschnagge** [comment]
introducing solver calculation into `models.h` is off-design. Consider moving move_path back into `solver.h`

**Comment by Necktschnagge** [comment]
detach solver from GUI thread,
check for heap delete effort, should also be moved into separate thread maybe

**Comment by Necktschnagge** [comment]
dont save state graphs for entire game history.
This is not a real memory leak, but behaves like one

**Comment by Necktschnagge** [comment]
! state safety is not guaranteed, please fix.

File: src/gui_helper.h [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+#include <QMessageBox>
+#include <QString>
+
+
+inline void showErrorDialog(const QString& message) {
+	QMessageBox msgBox(QMessageBox::Icon::Critical, QString("GUI ERROR"), message);
+	msgBox.exec();
```
Comment (Necktschnagge) [inline src/gui_helper.h:9]: Windows: it does not set focus on OK button. One can not just press ENTER to quit the error dialog.
@r0ost3r Can you check if it is the same behaviour on Linux?

File: src/models.h [hunk 2]
```diff
@@ -696,21 +728,139 @@ namespace tobor {
...
+
+						++color_next[i];
+					}
+				}
+
+				... to much effort in coding for a footprint. will also probably take longer computation time.
+			}
+			*/
+
+			inline move_path color_sorted_footprint() const {
```
Comment (Necktschnagge) [inline src/models.h:778]: unused

Check time consuption for state graph exploration versus equivalence class analysis.
Check optimization potential for equivalence class construction

**Comment by Necktschnagge** [comment]
We can store positions in 8bit integer when using 16x16 world.
But for computing size or comparing end iterators we need at least 9bit.

tobor_world should check for the correct integer size,
should also deliver a `smallest_position_int`, and an `int_type`

**Comment by Necktschnagge** [comment]
bug: state safety: run a solver when already started solver

**Comment by Necktschnagge** [comment]
Ideas of further engine improvement:
* collects statistics about optimal predecessor vector size depending on 

* priority queue for expanding using the rook-distance / min distance between target robot and goal.

* implement color-agnostic variant for GUI. Need to convert color-agnostic paths into colored paths.

* try to overcome big std::map for insert and search time...
   * is it possible to write all reached states of one expansion level into a vector, sort it afterwards?
   * map insert is not a good operation when looking at allocation.
      * passing a custom allocator can be a solution

* destruction of big containers has to be done in a separate thread. when solver data is thrown away (switching to a new game e.g.) dtors should not block UI

**Comment by Necktschnagge** [comment]
Add a feature for finding worlds of generators:
* input the world
* output the generator's counter for trivial group-generator 1

File: src/gui_helper.h (hunk 1 above)
Comment (Necktschnagge) [reply src/gui_helper.h:9]: It is not about the focus, but `grabKeyboard()` blocks **ENTER** for the button.
We need some `grab` - ON - OFF logic.

File: src/gui_interactive_controller.cpp [hunk 3]
```diff
@@ -4,85 +4,195 @@
...
+	//auto b = x % 4; x /= 4;
+	//auto c = x % 4; x /= 4;
+	//auto d = x % 4; x /= 4;
+	//auto e = x;
+	//qDebug() << a << "   :   " << b << "   :   " << c << "   :   " << d << "   :   " << e << "   :   " << std::get<1>(tup) << "   :   " << std::get<2>(tup) << "\n";
+	//qDebug() << originalGenerator.get_counter() << "\n";
+	//auto world = originalGenerator.get_tobor_world();
+	//auto target = originalGenerator.get_target_cell();
+
+	auto world = GameController::world_type(16, 16);
```
Comment (Necktschnagge) [inline src/gui_interactive_controller.cpp:30]: Create some *static-sized* board such that we can use 8bit for positions.

File: src/gui_interactive_controller.cpp [hunk 4]
```diff
@@ -105,35 +215,217 @@ void GuiInteractiveController::refreshSVG()
...
+		mainWindow->ui->actionMoveBack->setEnabled(false);
+
+	}
+	else if (interactive_mode == InteractiveMode::SOLVER_INTERACTIVE_STEPS) {
+
+		mainWindow->ui->actionNewGame->setEnabled(false);
+
+		mainWindow->ui->actionStopGame->setEnabled(true);
+
+		mainWindow->ui->actionStart_Solver->setEnabled(true);
```
Comment (Necktschnagge) [inline src/gui_interactive_controller.cpp:248]: nested solver not yet state-safe, deactivate this.

File: src/gui_interactive_controller.cpp [hunk 5]
```diff
@@ -105,35 +215,217 @@ void GuiInteractiveController::refreshSVG()
...
+	if (interactive_mode != InteractiveMode::SOLVER_INTERACTIVE_STEPS) {
+		return showErrorDialog("Cannot select any solver solution when not in solver mode!");
+	}
+	gameHistory.back().selectSolution(index);
+	gameHistory.back().resetSolverSteps();
+}
+
+void GuiInteractiveController::viewSolutionPaths()
+{
+	QStringList qStringList;
```
Comment (Necktschnagge) [inline src/gui_interactive_controller.cpp:332]: function has to be improved!

File: src/gui_interactive_controller.h [hunk 6]
```diff
@@ -110,38 +177,59 @@ class GuiInteractiveController final {
...
 public:
 
+	void startReferenceGame();
+
 	GuiInteractiveController(MainWindow* mainWindow) :
 		mainWindow(mainWindow),
 		interactive_mode(InteractiveMode::NO_GAME),
 		originalGenerator()
 	{
-		std::random_device rd;
```
Comment (Necktschnagge) [inline src/gui_interactive_controller.h:120]: make the using choosing for random initial board or not.

File: src/mainwindow.cpp [hunk 7]
```diff
@@ -145,22 +145,7 @@ void MainWindow::keyPressEvent(QKeyEvent* e)
...
-	list << "a" << "b" << "c";
-	list << "a" << "b" << "c";
-	list << "a" << "b" << "c";
-	list << "a" << "b" << "c";
-	list << "a" << "b" << "c";
-	list << "a" << "b" << "c";
-	model->setStringList(list);
-
-	ui->listView->setModel(model);
+	guiInteractiveController.startReferenceGame();
```
Comment (Necktschnagge) [inline src/mainwindow.cpp:148]: change name and location of the main menu button

File: src/mainwindow.h [hunk 8]
```diff
@@ -30,7 +30,7 @@ class MainWindow : public QMainWindow
 		std::unique_ptr<QGraphicsScene> q_graphics_scene;
 		/*
 		*/
-		inline SvgViewToolchain& operator =(SvgViewToolchain&& another) {
+		inline SvgViewToolchain& operator =(SvgViewToolchain&& another) noexcept {
```
Comment (Necktschnagge) [inline src/mainwindow.h:33]: check if noexcept is correct or needed!

File: src/models.h [hunk 9]
```diff
@@ -696,21 +739,269 @@ namespace tobor {
...
+				std::stable_sort(
+					result.vector().begin(),
+					result.vector().end(),
+					[](const piece_move_type& left, const piece_move_type& right) { return left.pid < right.pid; }
+				);
+
+				return result;
+			}
+
+			inline bool is_interleaving_neighbour(const move_path& another) const {
```
Comment (Necktschnagge) [inline src/models.h:835]: check for usage

File: src/models.h [hunk 10]
```diff
@@ -696,21 +739,269 @@ namespace tobor {
...
+								search_begin->second |= REACHED;
+								indices_to_explore.insert(search_begin - flagged_paths.begin());
+							}
+
+						}
+					}
+
+					remaining_end = std::remove_if(
+						flagged_paths.begin(),
+						remaining_end,
```
Comment (Necktschnagge) [inline src/models.h:935]: why was it not working with vector end instead of remaining end?

File: src/models.h [hunk 11]
```diff
@@ -696,21 +739,269 @@ namespace tobor {
...
+
+				return equivalence_classes;
+			}
+
+			inline static std::vector<std::vector<move_path>> interleaving_partitioning(const std::vector<move_path>& paths) {
+
+				static constexpr bool USE_IMPROVEMENT{ true };
+
+				if constexpr (USE_IMPROVEMENT) {
+					return interleaving_partitioning_improved(paths);
```
Comment (Necktschnagge) [inline src/models.h:955]: make this a feature toggle

**Comment by Necktschnagge** [comment]
Check all changes on `models.h` and `solver.h` again!

**Comment by Necktschnagge** [comment]
This PR is getting too big.

I want to merge this status quo as soon as we checked state-safety by some manual testing.

Engine changes have to be checked again by @Necktschnagge before merge.

@r0ost3r please review, and read the code carefully to identify bugs.
please checkout this branch and test if the solver is working, run through the app several times,
try to push buttons (or use the keyboard short-cuts) as many as you can, **especially push those that you think you should not push in that situation!**. E.g. when trying to start the solver while there is no game started, the app should not hang.

If you get annoying error message boxes, this is wanted behaviour for now.
We should get rid of them in following PRs. But for now they are supposed to prevent the user from running in undefined behaviour.

If you find undefined behaviour or running into ill application states, report this.
Any of this should be **treated as a blocker for this pull request**.

But the application should never stop responding, or

**Comment by r0ost3r** [comment]
👍🏼

**Review by r0ost3r — APPROVED** [review]

File: src/gui_interactive_controller.cpp [hunk 12]
```diff
@@ -1,88 +1,178 @@
 #include "predefined.h"
```
Comment (Necktschnagge) [inline src/gui_interactive_controller.cpp:1]: The whole is ensured to be reviewed again in #60

File: src/solver.h [hunk 13]
```diff
@@ -528,6 +531,46 @@ namespace tobor {
 
 			inline size_type get_optimal_solution_size() { return optimal_solution_size; };
 
+			template<class Insert_Iterator>
+			inline void optimal_path_helper_back_to_front(map_iterator state, Insert_Iterator destination, const move_path_type& rest_path = move_path_type()) {
```
Comment (Necktschnagge) [inline src/solver.h:535]: still need to review / improve this function

File: src/solver.h [hunk 14]
```diff
@@ -528,6 +531,46 @@ namespace tobor {
...
+					path.vector()[0] = move;
+					std::copy(rest_path.vector().cbegin(), rest_path.vector().cend(), path.vector().begin() + 1);
+
+					optimal_path_helper_back_to_front(predecessor_map_iterator, destination, path);
+				}
+
+
+			}
+
+			inline std::map<positions_of_pieces_type, std::vector<move_path_type>> optimal_paths(const cell_id_type& target_cell) {
```
Comment (Necktschnagge) [inline src/solver.h:554]: still need to review this function

File: src/solver.h [hunk 15]
```diff
@@ -528,6 +531,46 @@ namespace tobor {
...
+
+						optimal_path_helper_back_to_front(iter, std::back_inserter(result[state]));
+					}
+				}
+
+				return result;
+
+			}
 
 			inline void build_state_graph_for_all_optimal_solutions(
```
Comment (Necktschnagge) [inline src/solver.h:575]: still need to optimize this function

File: src/gui_helper.h (hunk 1 above)
Comment (Necktschnagge) [reply src/gui_helper.h:9]: moved to #78

File: src/models.h (hunk 2 above)
Comment (Necktschnagge) [reply src/models.h:823]: (#85 )

**Comment by Necktschnagge** [comment]
I created tons of new issues to check single points later.

**Pull request merged**